resdk/data_upload/annotate_samples.py

```diff
@@ -0,0 +1,141 @@
+"""Annotate samples in a collection."""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+from resdk.data_upload.samplesheet import FileExporter, FileImporter
+
+__all__ = ('annotate_samples', 'export_annotation')
+
+logger = logging.getLogger(__name__)
+
+
+def annotate_samples(collection, samplesheet_source, schema='sample'):
+    """Batch-annotate the samples in a collection.
+
+    This method applies annotations to the samples in a collection based on a
+    sample annotation spreadsheet. The entries in the spreadsheet and the
+    samples in the collection are matched by sample name, and validated for
+    upload.
+
+    :param str samplesheet_source: path to a local sample annotation
+        spreadsheet, or name of one in the collection, or an existing
+        FileImporter object created from an upload
+    :param str schema: slug of the descriptor schema to use
+    """
+    # Extract the information from a sample annotation spreadsheet
+    if isinstance(samplesheet_source, FileImporter):
+        samplesheet = samplesheet_source
+    else:
+        samplesheet = FileImporter(samplesheet_source)
+
+    # TODO: Try to pull the samplesheet as a data object from the collection
+    #       once it is possible to upload samplesheets
+    # else:
+    #     sheet = collection.data.get(name=samplesheet_source,
+    #                                 type='data:annotation')
+    #     sheet.download()
+    #     samplesheet = FileImporter(sheet.files()[0])
+    #     err_samples = []
+
+    # Report summary of invalid annotations
+    err_samples = samplesheet.invalid_samples
+    if err_samples:
+        logger.error(
+            "\nInvalid annotations were provided for the following samples: %s."
+            "\nPlease fill in all bolded columns of the template "
+            "generated by the `export_annotation` method of"
+            " your collection.",
+            ', '.join(err_samples)
+        )
+
+    # Match annotation entries to collection samples
+    matched = map(
+        lambda entry: (entry, collection.samples.filter(name=entry.name)),
+        samplesheet.sample_list
+    )
+
+    # Partition out the samples missing from the collection
+    coll_samples, missing_samples = _part_list(lambda pair: pair[1], matched)
+    err_missing = [
+        name for name in err_samples if not collection.samples.filter(name=name)
+    ]
+    missing_names = list(map(lambda pair: pair[0].name, missing_samples)) + err_missing
+
+    if missing_samples:
+        logger.error(
+            "\nThe following samples are missing from the collection:"
+            " %s. \nPlease check the names.",
+            ', '.join(missing_names)
+        )
+
+    # Apply each annotation to a sample in the collection
+    logger.info("\nAnnotating Samples...\n")
+    sample_count = 0
+    for entry, samples in coll_samples:
```

Comment:
This does not work because `map` and `filter` objects are lazy-loaded in Python. The object needs to be evaluated in order to apply the functions or extract any values. If I try to filter the same object in several different ways, as soon as one is evaluated, it interferes with all of the other instances. 

If I try to use `map` to apply the annotations, they are not applied until the object is explicitly evaluated as a list. But because we don't actually need a list of return values from that operation, that method has poorer performance and uses more memory than simply using a `for` loop.

Reply:
In general, when no return values are required, `for` loops [actually give the best performance](http://leadsift.com/loop-map-list-comprehension/). When return values are required, list comprehension is usually better than `map`. 

Though, I suppose this performance data would be different if we were using numpy arrays.

Reply:
I'm not concerned about performance of this `for` loop. Each iteration (sample annotation) takes orders of magnitude longer than the iteration time difference between the two implementations anyway. If we are concerned about memory footprint with big spreadsheets we should have implemented `FileImporter` as a generator in the first place.

I made this suggestion to improve code readability and possibly allow straightforward support for concurrency. But let's move on, @dblenkus will have the final say on this PR anyway.